go/controller/api/v1alpha1/observabilityprovider_types.go

```diff
@@ -0,0 +1,91 @@
+package v1alpha1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/pluralsh/console/go/client"
+)
+
+func init() {
+	SchemeBuilder.Register(&ObservabilityProvider{}, &ObservabilityProviderList{})
+}
+
+// ObservabilityProviderList contains a list of ObservabilityProvider.
+// +kubebuilder:object:root=true
+type ObservabilityProviderList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []ObservabilityProvider `json:"items"`
+}
+
+// ObservabilityProvider defines metrics provider used
+// by i.e. InfrastructureStack to determine if a stack run
+// should be cancelled.
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Namespaced
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="ID",type="string",JSONPath=".status.id",description="ID of the ObservabilityProvider in the Console API."
+type ObservabilityProvider struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +kubebuilder:validation:Required
+	Spec ObservabilityProviderSpec `json:"spec"`
+
+	// +kubebuilder:validation:Optional
+	Status Status `json:"status,omitempty"`
+}
+
+func (in *ObservabilityProvider) SetCondition(condition metav1.Condition) {
+	meta.SetStatusCondition(&in.Status.Conditions, condition)
+}
+
+func (in *ObservabilityProvider) Diff(hasher Hasher) (changed bool, sha string, err error) {
+	currentSha, err := hasher(in.Spec)
+	if err != nil {
+		return false, "", err
+	}
+
+	return !in.Status.IsSHAEqual(currentSha), currentSha, nil
+}
+
+func (in *ObservabilityProvider) Attributes(credentials client.ObservabilityProviderCredentialsAttributes) client.ObservabilityProviderAttributes {
+	return client.ObservabilityProviderAttributes{
+		Name:        in.Spec.Name,
+		Type:        in.Spec.Type,
+		Credentials: credentials,
+	}
+}
+
+type ObservabilityProviderSpec struct {
+	// Name of the ObservabilityProvider in the Console API.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// Type of the ObservabilityProvider.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum=DATADOG;NEWRELIC
+	Type client.ObservabilityProviderType `json:"type"`
+
+	// Credentials to access the configured provider Type.
+	// +kubebuilder:validation:Required
```

Comment:
These shouldn't be required ideally (If you already created in-ui)

Reply:
IIRC GraphQL schema required it, that's why I have made it also required.